clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -2769,6 +2769,24 @@ LogicalResult AtomicFetch::verify() {
   return mlir::success();
 }
 
+LogicalResult BinOp::verify() {
+  bool noWrap = getNoUnsignedWrap() || getNoSignedWrap();
+
+  if (!isa<mlir::cir::IntType>(getType()) && noWrap)
+    return emitError()
+           << "only operations on integer values may have nsw/nuw flags";
+
+  bool noWrapOps = getKind() == mlir::cir::BinOpKind::Add ||
+                   getKind() == mlir::cir::BinOpKind::Sub ||
+                   getKind() == mlir::cir::BinOpKind::Mul;
+
+  if (noWrap && !noWrapOps)
+    return emitError() << "The nsw/nuw flags are applicable to Add, Sub and "
+                          "Mul operations only";
```

Comment:
Thanks! Please add 'invalid.cir' tests and nit one the error message: `The nsw/nuw flags are applicable to opcodes: 'add', 'sub' and 'mul'`.